src/components/tl/mlx5/mcast/tl_mlx5_mcast_coll.c

```diff
@@ -1,22 +1,248 @@
+            if (req->first_send_psn + num_sent < comm->last_acked + wsize &&
+                req->to_send) {
+                /* How many to send: either all that are left (if they fit into window) or
+                   up to the window limit */
+                to_send = ucc_min(req->to_send,
+                                  comm->last_acked + wsize - (req->first_send_psn + num_sent));
+                ucc_tl_mlx5_mcast_send(comm, req, to_send, zcopy);
+
+                num_free_win = wsize - (comm->psn - comm->last_acked);
+            }
+        }
+        
+        status = ucc_tl_mlx5_mcast_prepare_reliable(comm, req, req->root);
+        if (ucc_unlikely(UCC_OK != status)) {
+            return status;
+        }
+
+        if (num_free_win && req->to_recv) {
+            /* How many to recv: either all that are left or up to the window limit. */
+            pending_q_size = 0;
+            to_recv        = ucc_min(num_free_win, req->to_recv);
+            to_recv_left   = ucc_tl_mlx5_mcast_recv(comm, req, to_recv, &pending_q_size);
+
+            if (to_recv == to_recv_left) {
+                /* We didn't receive anything: increase the stalled counter and get ready for
+                   drop event */
+                if (comm->stalled++ >= DROP_THRESHOLD) {
+
+                    tl_trace(comm->lib, "Did not receive the packet with psn in"
+                            " current window range, so get ready for drop"
+                            " event. pending_q_size %d current comm psn %d"
+                            " last_acked psn %d stall threshold %d ",
+                            pending_q_size, comm->psn, comm->last_acked,
+                            DROP_THRESHOLD);
+
+                    status = ucc_tl_mlx5_mcast_bcast_check_drop(comm, req);
+                    if (UCC_INPROGRESS == status) {
+                        return status;
+                    }
+                }
+            } else if (to_recv_left < 0) {
+                /* a failure happend during cq polling */
+                return UCC_ERR_NO_MESSAGE;
+            } else {
+                comm->stalled = 0;
+                comm->timer   = 0;
+            }
+        }
+
+        /* This function will check if we have to do a round of reliability protocol */
+        status = ucc_tl_mlx5_mcast_r_window_recycle(comm, req);
+        if (UCC_OK != status) {
+            return status;
+        }
+    }
+
+    if (req->to_send || req->to_recv || (zcopy && comm->psn != comm->last_acked)) {
+        return UCC_INPROGRESS;
+    } else {
+        return status;
+    }
+}
+
+
+ucc_status_t ucc_tl_mlx5_mcast_test(ucc_tl_mlx5_mcast_coll_req_t* req)
+{
+    ucc_status_t status = UCC_OK;
+
+    ucc_assert(req->comm->psn >= req->start_psn);
+
+    status = ucc_tl_mlx5_mcast_do_bcast(req);
+    if (UCC_OK == status) {
+        ucc_assert(req->comm->ctx != NULL);
+        ucc_tl_mlx5_mcast_mem_deregister(req->comm->ctx, req->rreg);
```

Comment:
the status of `ucc_tl_mlx5_mcast_mem_deregister` is not checked

Reply:
In fact, this can be converted to void return as the corresponding UCX call is void return.